packages/medusa/src/services/totals.js

```diff
@@ -0,0 +1,197 @@
+   * @return {int} the calculated subtotal
+   */
+  getSubTotal(cart) {
+    let subTotal = 0
+    if (!cart.items) {
+      return subTotal
+    }
+
+    cart.items.map(item => {
+      if (Array.isArray(item.content)) {
+        const temp = _.sumBy(item.content, c => c.unit_price * c.quantity)
+        subTotal += temp * item.quantity
+      } else {
+        subTotal +=
+          item.content.unit_price * item.content.quantity * item.quantity
+      }
+    })
+    return subTotal
+  }
+
+  /**
+   * Calculates either fixed or percentage discount of a variant
+   * @param {string} lineItem - id of line item
+   * @param {string} variant - id of variant in line item
+   * @param {int} variantPrice - price of the variant based on region
+   * @param {int} valye - discount value
+   * @param {string} discountType - the type of discount (fixed or percentage)
+   * @return {{ string, string, int }} triples of lineitem, variant and
+   *    applied discount
+   */
+  calculateDiscount_(lineItem, variant, variantPrice, value, discountType) {
+    if (discountType === "percentage") {
+      return {
+        lineItem,
+        variant,
+        amount: (variantPrice / 100) * value,
+      }
+    } else {
+      return {
+        lineItem,
+        variant,
+        amount: value >= variantPrice ? variantPrice : value,
+      }
+    }
+  }
+
+  /**
+   * If the discount_rule of a discount has allocation="item", then we need
+   * to calculate discount on each item in the cart. Furthermore, we need to
+   * make sure to only apply the discount on valid variants. And finally we
+   * return ether an array of percentages discounts or fixed discounts
+   * alongside the variant on which the discount was applied.
+   * @param {Discount} discount - the discount to which we do the calculation
+   * @param {Cart} cart - the cart to calculate discounts for
+   * @return {[{ string, string, int }]} array of triples of lineitem, variant
+   *    and applied discount
+   */
+  async getAllocationItemDiscounts(discount, cart) {
+    let discounts = []
+    for (const item of cart.items) {
+      discount.discount_rule.valid_for.forEach(async v => {
+        if (Array.isArray(item.content)) {
+          item.content.forEach(async c => {
+            if (c.variant._id === v) {
+              const variantRegionPrice = await this.productVariantService_.getRegionPrice(
+                v,
+                cart.region_id
+              )
+              discounts.push(
+                this.calculateDiscount_(
+                  item._id,
+                  v,
+                  variantRegionPrice,
+                  discount.discount_rule.value,
+                  discount.discount_rule.type
+                )
+              )
+            }
+          })
+        } else {
+          if (item.content.variant._id === v) {
+            const variantRegionPrice = await this.productVariantService_.getRegionPrice(
```

Comment:
use the LineItem `unit_price` instead. We may have to think about a "works on already discounted products" rule at some point, but for now we just assume that the unit_price is where we get the price from. This also allows you to remove the `ProductVariantService` dependency :)

Reply:
Would the `unit_price` not differ from, what region you are in? That's the reason for doing it in the first place.

Reply:
We can assume that `unit_price` is already given in the cart's currency & region price. We may have to do some more work in the CartService to enforce this. The idea with `unit_price` is that you can override the actual price being paid if it becomes necessary. For example, if you want to do a sale a "`medusa-plugin-sale`" could use the `unit_price` field to set the discounted price of the line item instead of the store operator changing the price on the variant.

Reply:
Alright - makes sense! I'll replace it.

Reply:
Take a look at `setRegion` in the `CartService`.

Reply:
`unit_price` is found on `item.content.unit_price` not `item.unit_price` as I accidentally indicated in the first comment :)